pkg/hubble/parser/threefour/parser.go

```diff
@@ -280,7 +287,29 @@ func (p *Parser) Decode(data []byte, decoded *pb.Flow) error {
 	decoded.ProxyPort = decodeProxyPort(dbg, tn)
 	decoded.Summary = summary
 
-	if p.correlateL3L4Policy && p.endpointGetter != nil {
+	var correlated bool
+	// If we have a valid cookie use it to add log info and correlate the flow.
+	if policyCookie != 0 && p.policyMetadataGetter != nil {
+		// If provided, look up policy log string based on the cookie we got. This will be more
+		// specific than the policy log string retrieved from the correlated policy. Also don't make
+		// this conditional on policy correlation being enabled as this is a simple map lookup.
+		cookie, ok := p.policyMetadataGetter.GetCookie(policyCookie)
+		if !ok {
+			p.log.Debug(
+				"no policy cookie found",
+				logfields.PolicyCookie, policyCookie,
+			)
+		} else {
+			decoded.PolicyLog = cookie.Logs
+			// Now if we have have a valid cookie use the labels and revision metadata to correlate.
+			if p.correlateL3L4Policy && p.endpointGetter != nil && cookie.Labels != "" {
+				correlation.CorrelatePolicyFromCookie(p.log, cookie, decoded)
+				correlated = true
+			}
+		}
+	}
+
+	if !correlated && p.correlateL3L4Policy && p.endpointGetter != nil {
 		correlation.CorrelatePolicy(p.log, p.endpointGetter, decoded)
 	}
```

Comment:
LGTM but nitpicking: can we extract that logic into a func dedicated to policy correlation so we avoid adding more code to the already-too-long Decode func please? :pray: